src/util/sighash.rs

```diff
@@ -0,0 +1,904 @@
+        let mut enc = SigHash::engine();
+        self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type)?;
+        Ok(SigHash::from_engine(enc))
+    }
+
+    fn common_cache(&mut self) -> &CommonCache {
+        if self.common_cache.is_none() {
+            let mut enc_prevouts = sha256::Hash::engine();
+            let mut enc_sequences = sha256::Hash::engine();
+            for txin in self.tx.input.iter() {
+                txin.previous_output
+                    .consensus_encode(&mut enc_prevouts)
+                    .unwrap();
+                txin.sequence.consensus_encode(&mut enc_sequences).unwrap();
+            }
+            let cache = CommonCache {
+                prevouts: sha256::Hash::from_engine(enc_prevouts),
+                sequences: sha256::Hash::from_engine(enc_sequences),
+                outputs: {
+                    let mut enc = sha256::Hash::engine();
+                    for txout in self.tx.output.iter() {
+                        txout.consensus_encode(&mut enc).unwrap();
+                    }
+                    sha256::Hash::from_engine(enc)
+                },
+            };
+            self.common_cache = Some(cache);
+        }
+        self.common_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
+    }
+
+    fn segwit_cache(&mut self) -> &SegwitCache {
+        if self.segwit_cache.is_none() {
+            let cache = SegwitCache {
+                prevouts: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().prevouts).into_inner(),
+                ),
+                sequences: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().sequences).into_inner(),
+                ),
+                outputs: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().outputs).into_inner(),
+                ),
+            };
+            self.segwit_cache = Some(cache);
+        }
+        self.segwit_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
+    }
+
+    fn taproot_cache(&mut self, prevouts: &[TxOut]) -> &TaprootCache {
+        if self.taproot_cache.is_none() {
+            let mut enc_amounts = sha256::Hash::engine();
+            let mut enc_script_pubkeys = sha256::Hash::engine();
+            for prevout in prevouts {
+                prevout.value.consensus_encode(&mut enc_amounts).unwrap();
+                prevout
+                    .script_pubkey
+                    .consensus_encode(&mut enc_script_pubkeys)
+                    .unwrap();
+            }
+            let cache = TaprootCache {
+                amounts: sha256::Hash::from_engine(enc_amounts),
+                script_pubkeys: sha256::Hash::from_engine(enc_script_pubkeys),
+            };
+            self.taproot_cache = Some(cache);
+        }
+        self.taproot_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
```

Comment:
I'd like to avoid unwraps when possible: for instance, somebody may change the logic in the lines above and this unwrap will suddenly start panicking.

So something like `match self.taproot_cache { None => { ... }, Some(tc) => tc }` will be much more preferable

Reply:
I was going to write this is impossible (I previously didn't write the exact comment you did because I believed so) but decided to double-check. I found [`Option::get_or_insert_with()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with), which is perfect for this case and I believe it should be used.

Reply:
The reasons I didn't use `get_or_insert_with` are here https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r670262033
If we agree on that, I should surely add a comment because it's already been suggested https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r671297559

Reply:
Heh, missed those. If my [suggested](https://github.com/RCasatta/rust-bitcoin/pull/3) solution to remove unwraps is not accepted then comments should definitely be written.

Reply:
Suggested solution accepted, thanks